### PR TITLE
Make JSON symbol table input and output consistent

### PR DESCRIPTION
**Pull request by danpoe**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f88d91f6acde1d3709ee7dbda587b087efd97f2d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96648090

File: unit/json_symbol_table.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+  json_stream_arrayt json_stream_array;
+};
+
+void get_goto_model(std::istream &in, goto_modelt &goto_model)
+{
+  optionst options;
+  cbmc_parse_optionst::set_default_options(options);
+
+  null_message_handlert null_message_handler;
```
Comment (tautschnig) [inline unit/json_symbol_table.cpp]: You might want to use the one from unit/testing-utils/message.h.

**Review by tautschnig — APPROVED** [review]
Looks good to me, but should likely receive a review from someone who depends on the JSON symbol table.

**Review by peterschrammel — APPROVED** [review]

File: src/util/json_irep.cpp [hunk 2]
```diff
@@ -97,27 +96,54 @@ void json_irept::convert_named_sub_tree(
...
+
+    if(it != json_object.end())
+    {
+      for(const auto &named_sub : to_json_object(it->second))
+        out.add(named_sub.first) = convert_from_json(named_sub.second);
+    }
+  }
+
+  {
+    const auto it = json_object.find("comment");
```
Comment (smowton) [inline src/util/json_irep.cpp]: Not any more

File: src/util/symbol.cpp [hunk 3]
```diff
@@ -144,3 +144,22 @@ bool symbolt::is_well_formed() const
 
   return true;
 }
+
+bool symbolt::operator==(const symbolt &other) const
+{
+  return type == other.type && value == other.value &&
```
Comment (smowton) [inline src/util/symbol.cpp]: Suggest for readability do `// clang-format off` and list one condition per line

File: src/util/symbol_table.cpp [hunk 4]
```diff
@@ -219,3 +221,46 @@ void symbol_tablet::validate(const validation_modet vm) const
...
+{
+  // we cannot use == for comparing the multimaps as it compares the items
+  // sequentially, but the order of items with equal keys depends on the
+  // insertion order
+
+  {
+    std::vector<std::pair<irep_idt, irep_idt>> v1;
+    std::vector<std::pair<irep_idt, irep_idt>> v2;
+
+    for(const auto &pair : internal_symbol_base_map)
```
Comment (smowton) [inline src/util/symbol_table.cpp]: `v1(internal_symbol_base_map.begin(), internal_symbol_base_map.end())`, similarly `v2`.

**Review by chrisr-diffblue — APPROVED** [review]
We'll need to update gnat2goto, but we can do that when we next bump its cbmc submodule.

**Comment by danpoe** [comment]
@smowton Ok, addressed the comments.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5996247b097b11960f3f8e74dadf29df59fcefe2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96741193
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8d4ca27e26c7dba8afe4e73f4b3bb1cf30836305).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96741962

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: a3f38e711a177315e47a18cd7b5c374d8857e98a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96747712
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**